### ISSUE TITLE
New class RecentTrailers added

### ISSUE THREAD

**Issue by duck7000** [duck7000/imdbGraphQLPHP#81]
I added this class as there once was a request for it but never did add it.

It will get the (max 100 items) the latest added trailers from IMDb as seen on https://www.imdb.com/trailers/
Only latest added trailers are available so not the trending, popular or anticipated.

I added it just in case somebody would want to use it and because i did find out how to do it hah.

Check it out if you like, let me know if there any issues or requests with it.

**Comment by duck7000** [comment]
So far i have only found contentType but it only list trailer, nothing else. So i don't know where imdb gets official trailer from. I'll try again.
imdbid i'll try to find

**Comment by mosource21** [comment]
Is there a method to try to find what the field names are called or is it purely guess work?

**Comment by duck7000** [comment]
the fieldnames are only findable in chrome extension called GraphiQL but it is not easy. I have set it up with guzzlehttp and php developserver, that way i can get documentation about the whole structure. But again it is not easy to set up and use

**Comment by duck7000** [comment]
But the good news is that i have found what you are looking for!
I added:
contentType (this will get Official Trailer, Official Teaser etc
titleId (this will get the title id so you can construct the url to that title

It is in latest git commit, and i adjusted the wiki to match the new fields

**Comment by mosource21** [comment]
Brilliant 👍 Thanks

I will report back with more feedback.

**Comment by duck7000** [comment]
I fixed 2 major screw ups in this class..

1 the thumbnail image should not be from the trailer but the title it belongs to, fixed that.
2 Thumbnail itself is now 140x207 instead of the full image. I added the image processing functions to correctly get perfect thumbnails in the right size. I made the size fixed, but if anyone wants it can also be in config, let me know if somebody does want that.

**Comment by duck7000** [comment]
Well i might as well add recentVideo and trendingVideo together in a Trailers class?
I just thought of that, i thought trailers would become only 1 method but now i got 2 so this might be a good solution

**Comment by mosource21** [comment]
You are working quicker than I can keep up 😄 

Here is my feedback so far I was waiting until I had time to investigate further.  

1. (You have already mentioned this) Instead of calling RecentTrailers.php perhaps using a generic name Videos.php class would allow for future expansion.  A combined Trailers.php class is perfect and better describes current functionality than Videos.php class.
2. There is no support for cache/log - I think this needs __construct adding to Trailers.php class but I am not sure I haven't had time to investigate further.

As previously noted I will be having a proper look at this soon and providing feedback.

Thanks

**Comment by duck7000** [comment]
> You are working quicker than I can keep up 😄

Haha i'm also amazed how quick i come up with new stuff, i test it to work but later on i usually find this kind of issues.
So for production use release version as they are more stable.

I will combine those 2 methods in a Video class to keep it together and easier to maintain.

Cache or logging is in my view not needed as this is a kind of search class so caching results is not very useful as they change with almost every call?
Chart class is also not cached, just as TitleSearch or TitleSearchAdvanced
Unless you can provide very good reasons to do this?

**Comment by mosource21** [comment]
> Haha i'm also amazed how quick i come up with new stuff

I just don't like leaving feedback without at least attempting to suggest a solution and that can delay me quite a bit.

> I will combine those 2 methods in a Video class to keep it together and easier to maintain.

Do not feel you need to use Video.php unless you think there will be future plans for other things other than Trailers otherwise Trailers.php is perfectly fine.  Thinking about it further perhaps Trailers.php, WhatToWatch.php, IMDbOriginals.php etc would work better instead of a global Videos.php?  I would only use Trailers.php but your options are still open then.

> Cache or logging is in my view not needed as this is a kind of search class so caching results is not very useful as they change with almost every call? Chart class is also not cached, just as TitleSearch or TitleSearchAdvanced Unless you can provide very good reasons to do this?

I wouldn't class the Trailers.php class as a Search though.  It can stay the same for hours on end and having the cache also makes it much quicker when working with the class as continued GraphQL queries are not being made.

I will look at adding it hopefully it is only one line and I do not think it impacts other users if you do not want to use it but the functionality is there if you do.

Thanks

**Comment by duck7000** [comment]
I think i go with Trailers.php, it is nice common and, if i can, extend it with other video related methods.
The ones you mentioned are not (and possible can't get them) so they belong in their own class file.

Caching might indeed be handy as you mentioned, so i'll try to get that in.

**Comment by mosource21** [comment]
I am hoping this will work.  I do not mind giving it try though if you want.  Also including $config is a good idea also.

public function __construct(Config $config = null, LoggerInterface $logger = null, CacheInterface $cache = null)
 {
     parent::__construct($config, $logger, $cache);
     $this->imageFunctions = new Image();
 }

**Comment by duck7000** [comment]
hah i just did the same, but thanks for thinking with me though!

**Comment by mosource21** [comment]
I will give this a try and report back.  Not sure but can you use the config options for the hard coded 140/207 values in Trailer.php?

I have noticed the welcome return of config.php I have some suggestions for at least one addition but it can wait and I will open separate issue.

Thanks

**Comment by duck7000** [comment]
I deliberately choose not to move thumbnail width and height to config as the config is quickly filling up with settings that i need to properly explain what they are for.
Chart and TitleCombined doesn't have that either.

It can be added of course if you want that?

**Comment by duck7000** [comment]
I added thumbnail settings to config, setting is for all methods the same.
so if there will be more methods in Trailers class they will use the same thumbnail setting as well

**Comment by duck7000** [comment]
Wiki page is updated, All info is now in one wiki page

**Comment by mosource21** [comment]
> Check it out if you like, let me know if there any issues with it.

Sorry for delay but I have now spent quite some time trying this and apart from a few queries below everything is looking great (including caching, recent & trending).  

Many thanks 👍 

I have not really touched upon the images/thumbnails side of things though.  

> I added thumbnail settings to config, setting is for all methods the same.
> config is quickly filling up with settings that i need to properly explain

Consistency is an important factor and ensuring these finer details are taken care of so everything works the same makes the library look more polished/professional in my opinion.  

Config.php is getting quite comment heavy (as you have mentioned) but with the flexibility it provides I believe it is a very small "price" to pay!  

It has no impact on new users as they do not need to make any changes but the flexibility is there when you come more familiar with the library.

> Check it out if you like, let me know if there any requests with it.

* var_dump() has been accidentally left in (Title.php line 779)
* Rather than the movie release date is it possible to find and include the trailer published date?
* Rather than the runtime of the movie is it possible to find and include the trailer duration?
* Do you know what [OV] means that is sometimes included in the content type for example "Trailer [OV]"?

Thanks again

**Comment by duck7000** [comment]
As for your requests

- var_dump is removed (left over debug hah)
- In recentVideo() i added trailerCreateDate, this is a iso datestamp so it includes time as well. In trendingVideo this is not possible
- Trailer duration is already included as runtime but maybe i should rename that?
- I have no idea what that means, i'm from the Netherlands so i'm not that familiar with those terms

**Comment by duck7000** [comment]
> It is a pity something suitable is not available in trendingVideo though. Definitely in the realms of I do not know what I am doing but in the trendingTrailer GraphQL could the createdDate be in the latestTrailer{} bit along with the id/runtime & name?

Nope, not possible. There is a difference between videos or titles in query, videos have createdDate where titles does not.

I reverted the trailerRuntime part back to seconds in latest commit. I just thought it would make sense because the title runtime is also in minutes but you are right that for a (short) trailer seconds is better.

**Comment by mosource21** [comment]
Excellent, all good and working well 👍 

If you are happy and want to close the issue fine with me.

**Comment by duck7000** [comment]
Nice, thanks for testing and your comments!
